src/economy/economy.cpp

```diff
@@ -13,33 +13,591 @@
+	});
 
-	state.world.nation_get_gdp(n) -= amount * price * sat;
+	return total / count;
 }
 
-// it's registered as a demand separately
-void register_construction_demand(sys::state& state, dcon::nation_id n, dcon::commodity_id commodity_type, float amount) {
-	state.world.nation_get_construction_demand(n, commodity_type) += amount;
+float export_volume(
+	sys::state& state,
+	dcon::market_id s,
+	dcon::commodity_id c
+){
+	return state.world.market_get_export(s, c);
+}
+float export_volume(
+	sys::state& state,
+	dcon::nation_id s,
+	dcon::commodity_id c
+){
+	float total = 0.f;
+	state.world.nation_for_each_state_ownership(s, [&](auto soid) {
+		auto sid = state.world.state_ownership_get_state(soid);
+		auto market = state.world.state_instance_get_market_from_local_market(sid);
+		total += state.world.market_get_export(market, c);
+	});
+	return total;
 }
 
-void register_domestic_supply(sys::state& state, dcon::nation_id n, dcon::commodity_id commodity_type, float amount, economy_reason reason) {
-	state.world.nation_get_domestic_market_pool(n, commodity_type) += amount;
-	state.world.nation_get_gdp(n) += amount * state.world.commodity_get_current_price(commodity_type);
+
+trade_volume_data_detailed export_volume_detailed(
+	sys::state& state,
+	dcon::nation_id s,
+	dcon::commodity_id c
+) {
+	static auto per_nation_data = state.world.nation_make_vectorizable_float_buffer();
+
+	state.world.execute_serial_over_nation([&](auto nids) {
+		per_nation_data.set(nids, 0.f);
+	});
+
+	float total = 0.f;
+
+	state.world.nation_for_each_state_ownership(s, [&](auto soid) {
+		auto sid = state.world.state_ownership_get_state(soid);
+		auto market = state.world.state_instance_get_market_from_local_market(sid);
+		state.world.market_for_each_trade_route(market, [&](auto trade_route) {
+			auto current_volume = state.world.trade_route_get_volume(trade_route, c);
+			auto origin =
+				current_volume > 0.f
+				? state.world.trade_route_get_connected_markets(trade_route, 0)
+				: state.world.trade_route_get_connected_markets(trade_route, 1);
+			auto target =
+				current_volume <= 0.f
+				? state.world.trade_route_get_connected_markets(trade_route, 0)
+				: state.world.trade_route_get_connected_markets(trade_route, 1);
+
+			if(origin != market) return;
+
+			auto sat = state.world.market_get_direct_demand_satisfaction(origin, c);
+			auto absolute_volume = std::abs(current_volume);
+			auto s_origin = state.world.market_get_zone_from_local_market(origin);
+			auto s_target = state.world.market_get_zone_from_local_market(target);
+			auto n_origin = state.world.state_instance_get_nation_from_state_ownership(s_origin);
+			auto n_target = state.world.state_instance_get_nation_from_state_ownership(s_target);
+
+			auto distance = state.world.trade_route_get_distance(trade_route);
+
+			auto trade_good_loss_mult = std::max(0.f, 1.f - 0.0001f * distance);
+
+			if(n_origin != n_target) {
+				per_nation_data.get(n_target) += sat * absolute_volume;
```

Comment:
So, the issue here is that if this function is running in parallel with the game and the number of nations can grow, it is possible for a new nation to be created and found in the target list that has a greater id than what the buffer created above was sized to accommodate. If the number of nations can change, this function would have to check that the target id is within the size allocated in the buffer and simply discard it if it is not.